### PR TITLE
Networking: egress IP multi-NIC proposal

### PR DESCRIPTION
**Pull request by martinkennelly**

### CODE REVIEW

File: enhancements/network/egress-ip-multi-nic.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+---
+# Egress IP Multi-NIC
+## Summary
+For OpenShift release version 4.13 or less, users may utilize egress IP to allow source NAT when packets leave the primary network.
+This enhancement will add egress IP support for networks that exists on NICs that aren't managed by OVN and hence forth
+these networks are called secondary networks.
+
+## Motivation
+Users may not wish to route workload traffic over OVN managed primary network and instead prefer to use a secondary network.
+With this enhancement, we can support segregation of workload traffic over multiple NICs while allowing a consistent source IP address.
```
Comment (flavio-fernandes) [inline enhancements/network/egress-ip-multi-nic.md:28]: nit:  s / consistent / configurable

File: enhancements/network/egress-ip-multi-nic.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+## Proposal
+
+### Implementation Details/Notes/Constraints
+Users may provide not only a primary network but also additional networks that they want to utilize and support egress IP.
+Cluster manager will continue to provision `EgressIP` to nodes. Currently, it only assigns them to nodes which have a
+well-known label egress assignable `k8s.ovn.org/egress-assignable`.
+
+This enhancement is proposing enhancing cluster manager if the egress IP is within a subnet of a known non-primary network, then it
+will also consider which node hosts this secondary network. If no secondary network is found, we continue the default
+behaviour of assigning the egress IP to the primary network like what we do today.
```
Comment (flavio-fernandes) [inline enhancements/network/egress-ip-multi-nic.md:49]: This is likely a misconfiguration, but it may be worth mentioning here:
the eip will be potentially on the primary network of all nodes, or on the secondary. We do not antecipate the eip being on either one across nodes. ---  or it that a non-issue and we would be supporting whichever, on a node to node basis?

File: enhancements/network/egress-ip-multi-nic.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+We will also add a masquerade rule to iptables when a packet matching the source pod IP and also is leaving
+a particular interface.
+
+
+See section `Design Details` for full implementation details.
+
+### User Stories
+As a user, I want the ability for cluster workload traffic to be routed to a network that is not managed by OVN and to have a consist source IP
+that is configurable.
```
Comment (flavio-fernandes) [inline enhancements/network/egress-ip-multi-nic.md:75]: nit: ... and to have a configurable source IP.

File: enhancements/network/egress-ip-multi-nic.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+- Remove pod
+- Delete a valid rule manually 
+- Delete a valid IPtables rule manually
+- Delete a valid route manually
+- Delete a valid egress IP assigned to an interface manually
+- Valid egress service and egress IP functionality if both are active on the same k8 objs.
+Today, if egress service and egress IP both 'match' on the same pod, egress service is preferred.
+Continue this for this enhancement where egress service and egress IP overlap, prefer egress service.
+
+TBD: Should we have to consider, and therefore test NICs hot plugging / removal?
```
Comment (flavio-fernandes) [inline enhancements/network/egress-ip-multi-nic.md:96]: maybe not hot plugging, but maybe changing / creating / deleting secondary networks that would affect the eip validation. Including changes of mask size that affect longest prefix match

File: enhancements/network/egress-ip-multi-nic.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+- "@trozet"
+creation-date: 2023-05-16
+last-updated: 2023-05-30
+status: implementable
+tracking-link:
+- https://issues.redhat.com/browse/SDN-2097
+---
+# Egress IP Multi-NIC
+## Summary
+For OpenShift release version 4.13 or less, users may utilize egress IP to allow source NAT when packets leave the primary network.
```
Comment (flavio-fernandes) [inline enhancements/network/egress-ip-multi-nic.md:22]: 4.14 or less?

**Review by flavio-fernandes — COMMENTED** [review]
this is good stuff @martinkennelly !
some small-ish comments; please take hem as as suggestions and ignore what adds no value.

File: enhancements/network/egress-ip-multi-nic.md (hunk 2 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:49]: I removed this section - if `EgressIP` after learning today we make sure it falls within the primary network subnet. So, if egress IP doesnt find a network, we dont assign it to a node.

File: enhancements/network/egress-ip-multi-nic.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+- What about secondary bridges managed by OVN aka br-ex1
+  We ignore them and do not support egress IP.
+- Selecting the wrong network types from interfaces
+  Only consider addresses with scope `Global`.
+
+[1] https://en.wikipedia.org/wiki/Longest_prefix_match
+
+## Design Details
+### Cluster Manager
+If egress IP is within a subnet of a secondary network, it will need to assign this egress IP to a node which hosts this network.
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:147]: >I actually expect multiple nodes to have interfaces on the same subnet; how will it choose a node from the available ones ?

We expose all the networks we find on a node and annotate the `host-addresses` annotation including the networks mask.  
Cluster manager then decides which node contains a network which can "host" the egress IP.
 We use longest prefix match to decide which is the optimum network to choose. If there are multiple nodes that match, then one of them will be chosen at random as the host to host the egress IP.

[the remark this answers is not in the document]
Comment (maiqueb) [reply enhancements/network/egress-ip-multi-nic.md:147]: OK, that's an option; thanks for explaining.

If possible, I would add this info to the doc. 

Would it make sense to have some sort of algorithm that ensures the number of egress IPs across nodes are on par instead of going "full" random ?

File: enhancements/network/egress-ip-multi-nic.md [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+last-updated: 2023-05-30
+status: implementable
+tracking-link:
+- https://issues.redhat.com/browse/SDN-2097
+---
+# Egress IP Multi-NIC
+## Summary
+For OpenShift release version 4.13 or less, users may utilize egress IP to allow source NAT when packets leave the primary network.
+This enhancement will add egress IP support for networks that exists on NICs that aren't managed by OVN and hence forth
+these networks are called secondary networks.
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:24]: Changed it.

File: enhancements/network/egress-ip-multi-nic.md [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+## Motivation
+Users may not wish to route workload traffic over OVN managed primary network and instead prefer to use a secondary network.
+With this enhancement, we can support segregation of workload traffic over multiple NICs while allowing a configurable source IP address.
+
+### Goals
+- Support egress IP for secondary networks
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:31]: changed it

File: enhancements/network/egress-ip-multi-nic.md [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+### Non-Goals
+- Support egress IP on non-default networks controllers aka secondary network controllers
+- Support cloud egress IP assignment for secondary interfaces using cloud network config controller
+  - This maybe easy to do eventually, but it will require a lot of extra testing, so excluding it for now, and it's
+    a stretch goal.
+- Support egress IP on secondary bridges managed by OVN aka br-ex1
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:38]: Its ICNI bridge aka `br-ex1`

File: enhancements/network/egress-ip-multi-nic.md [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+- "@trozet"
+creation-date: 2023-05-16
+last-updated: 2023-05-30
+status: implementable
+tracking-link:
+- https://issues.redhat.com/browse/SDN-2097
+---
+# Egress IP Multi-NIC
+## Summary
+For OpenShift release version 4.13 or less, users may utilize egress IP to allow source NAT when packets leave the primary network.
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:22]: done

File: enhancements/network/egress-ip-multi-nic.md [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+```golang
+// The per node status, for those egress IPs who have been assigned.
+type EgressIPStatusItem struct {
+// Assigned node name
+Node string `json:"node"`
+// Assigned egress IP
+EgressIP string `json:"egressIP"`
+// Route via host network or OVN
+RouteViaHost bool `json:"routeViaHost"`
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:106]: yes, updated it to be clearer.

File: enhancements/network/egress-ip-multi-nic.md [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+This enhancement is proposing enhancing cluster manager if the egress IP is within a subnet of a known non-primary network, then it
+will also consider which node hosts this secondary network.
+
+Cluster manager will therefore need to understand all available networks within the cluster and this enhancement proposes
+to add all secondary interfaces addresses and subnet mask to existing annotation `k8s.ovn.org/host-addresses` which is
+added to all node custom resource objects.
+
+Egress IP must not equal an address already assigned to an interface, and we can use the aforementioned `k8s.ovn.org/host-addresses`
+annotation to detect this scenario.
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:55]: Yeah, its up to the user to make sure this is the case. I am saying here that we just enforce that the IP is not already assigned to any interface through the kubernetes cluster. Cluster manager would reject assigning the egress IP if it matched any IP seen within the cluster via `host-addresses` annotation.

File: enhancements/network/egress-ip-multi-nic.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:147]: Added to the doc.
yes, I need to revisit this algorithm with these changes and ensure we spread the load.

**Comment by martinkennelly** [comment]
>I also wonder what happens in there's a follow-up day2 change; let's say my nodes are connected to 2 non-OVN managed networks.
t=0: configure egress IPs for one of these non-ovn managed networks
t=1: use nmstate operator to configure a bond with those 2 interfaces
What would happen ? Do you want to support this scenario ? Is there a way for it to be prevented if you're not planning on supporting it ?

I hope this would be supported I think because the IP would move to the bond right? and the former interface is now a slave to the bond. During a reconcile, it would assign the egress IP to the bond and clean up any left over artifacts.

Ill add a test for this thank you as always Miguel.

**Comment by maiqueb** [comment]
> > I also wonder what happens in there's a follow-up day2 change; let's say my nodes are connected to 2 non-OVN managed networks.
> > t=0: configure egress IPs for one of these non-ovn managed networks
> > t=1: use nmstate operator to configure a bond with those 2 interfaces
> > What would happen ? Do you want to support this scenario ? Is there a way for it to be prevented if you're not planning on supporting it ?
> 
> I hope this would be supported I think because the IP would move to the bond right? and the former interface is now a slave to the bond. During a reconcile, it would assign the egress IP to the bond and clean up any left over artifacts.
> 
> Ill add a test for this thank you as always Miguel.

Yeah, afaiu, your reconcile loop should delete the old rule and create a new one pointing at the bond.

File: enhancements/network/egress-ip-multi-nic.md [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+- Remove pod
+- Delete a valid rule manually 
+- Delete a valid IPtables rule manually
+- Delete a valid route manually
+- Delete a valid egress IP assigned to an interface manually
+- Valid egress service and egress IP functionality if both are active on the same k8 objs.
+Today, if egress service and egress IP both 'match' on the same pod, egress service is preferred.
+Continue this for this enhancement where egress service and egress IP overlap, prefer egress service.
+- Egress IP interface is now a slave to a bond which is now the IP of the EIP interface. Reconcile to move the EIP to the bond.
+- Egress IP interface is now a slave to a bond which is now a different IP of the original EIP interface. Reconcile to move the EIP
```
Comment (maiqueb) [inline enhancements/network/egress-ip-multi-nic.md:147]: let's replace `slave` with something more inclusive.

member of a bond ?... I honestly don't know what's used now in the kernel.

File: enhancements/network/egress-ip-multi-nic.md [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+Cluster manager will therefore need to understand all available networks within the cluster and this enhancement proposes
+to add all non-ovn managed interfaces addresses and subnet mask to existing annotation `k8s.ovn.org/host-addresses` which is
+added to all node custom resource objects.
+
+Egress IP must not equal an address already assigned to an interface, and we can use the aforementioned `k8s.ovn.org/host-addresses`
+annotation to detect this scenario. The addresses that can be added into this annotation are filtered by the following criteria:
+- Link must be up
+- Link must not be loopback, type openvswitch or bridge
+- Link must not have a master index (aka bridge)
+- Link must not have an associated parent index (aka slave devices)
```
Comment (maiqueb) [inline enhancements/network/egress-ip-multi-nic.md:61]: I'd remove this master / slave terminology for something more inclusive.

**Review by maiqueb — COMMENTED** [review]
Looking good.

**Comment by martinkennelly** [comment]
Patryk may or may not review this. I think its well understood anyway and Tims validation is all thats needed IMO.

File: enhancements/network/egress-ip-multi-nic.md (hunk 14 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md]: Done. I cannot remove 'master' because thats what netlink calls it. I removed the slave reference in favour of child. Thanks!

File: enhancements/network/egress-ip-multi-nic.md (hunk 13 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:147]: changed it to child

File: enhancements/network/egress-ip-multi-nic.md [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+// Route via an OVN managed network or via a non OVN managed network
+RouteViaHost bool `json:"routeViaHost"`
+}
+```
+
+This will aid debugging of egress IP assignments but also aid egress IP controller located within ovnkube node to not
+have to recalculate if the egress IP is primary or secondary networks.
+
+### Risks and Mitigations
+- OVN kube node must recalculate all the matched pods for a given egress IP. This will require refactoring this logic
```
Comment (kyrtapz) [inline enhancements/network/egress-ip-multi-nic.md:169]: I am confused with this sentence, cluster manager controller for egress IP doesn't care about pods?

**Comment by martinkennelly** [comment]
Forced push to remove `routeViaHost` bool and changed it to `network` string. This was pointed out by @kyrtapz due to restriction on using bool type fields.

File: enhancements/network/egress-ip-multi-nic.md (hunk 15 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:168]: fixed.

File: enhancements/network/egress-ip-multi-nic.md [hunk 16]
```diff
@@ -0,0 +1,278 @@
...
+
+## Design Details
+### Cluster Manager
+If egress IP is within a subnet of a non-ovn managed network, it will need to assign this egress IP to a node which hosts this network.
+
+### OVN kube controller (aka ovn kube master, aka network controller manager)
+Enhance OVN logical router `ovn_cluster_router` to make the next-hop choice a little smarter. If the egress IP is assigned
+to a secondary network, then reroute it the management port of the node that hosts the secondary network.
+
+### OVN kube node
```
Comment (kyrtapz) [inline enhancements/network/egress-ip-multi-nic.md:207]: What will happen on ovnkube-node when multiple secondary egressIPs assigned to it match the same pod?

**Review by kyrtapz — APPROVED** [review]
Overall looks good to me, left some minor comments

File: enhancements/network/egress-ip-multi-nic.md [hunk 17]
```diff
@@ -0,0 +1,278 @@
...
+Cluster manager would assign the `EgressIP` to worker 2.
+
+To allow packets to be steered into secondary networks, OVN reroute logical policies on `ovn_cluster_router` will redirect
+the packet out to the management port of the node which the egress IP is hosted.
+
+We will add rules in linux routing policy database control for each source Pod IP and lookup the correct routing table
+which will contain one route. This single route within a new managed routing table will direct the packet to the correct interface.
+
+We will also add a masquerade rule to iptables when a packet leaving a particular interface matches the source pod IP.
+
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:120]: yes

File: enhancements/network/egress-ip-multi-nic.md [hunk 18]
```diff
@@ -0,0 +1,278 @@
...
+This enhancement is proposing enhancing cluster manager if the egress IP is within a subnet of a known non-primary network, then it
+will also consider which node hosts this secondary network.
+
+Cluster manager will therefore need to understand all available networks within the cluster and this enhancement proposes
+to add all non-ovn managed interfaces addresses and subnet mask to existing annotation `k8s.ovn.org/host-addresses` which is
+added to all node custom resource objects.
+
+Egress IP must not equal an address already assigned to an interface, and we can use the aforementioned `k8s.ovn.org/host-addresses`
+annotation to detect this scenario. The addresses that can be added into this annotation are filtered by the following criteria:
+- Link must be up
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:58]: reconcile will remove it and try to reassign. if it cannot reassign, it will clear the status.

File: enhancements/network/egress-ip-multi-nic.md [hunk 19]
```diff
@@ -0,0 +1,278 @@
...
+  Only consider addresses with scope `Global`.
+
+[1] https://en.wikipedia.org/wiki/Longest_prefix_match
+
+## Design Details
+### Cluster Manager
+If egress IP is within a subnet of a non-ovn managed network, it will need to assign this egress IP to a node which hosts this network.
+
+### OVN kube controller (aka ovn kube master, aka network controller manager)
+Enhance OVN logical router `ovn_cluster_router` to make the next-hop choice a little smarter. If the egress IP is assigned
```
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:204]: yes, that should work. I will add it as a test to the enhancement.

File: enhancements/network/egress-ip-multi-nic.md (hunk 16 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:209]: Undefined right? Today, what happens is undefined.

File: enhancements/network/egress-ip-multi-nic.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:58]: I added this as a test case.

File: enhancements/network/egress-ip-multi-nic.md (hunk 16 above)
Comment (kyrtapz) [reply enhancements/network/egress-ip-multi-nic.md:213]: I think it is fair to state at the enhancement level that it is undefined, we can see if this can be addressed in the  implementation

Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:213]: Done

File: enhancements/network/egress-ip-multi-nic.md [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+
+```golang
+// The per node status, for those egress IPs who have been assigned.
+type EgressIPStatusItem struct {
+// Assigned node name
+Node string `json:"node"`
+// Assigned egress IP
+EgressIP string `json:"egressIP"`
+// Assigned network
+Network string `json:"network"`
```
Comment (trozet) [inline enhancements/network/egress-ip-multi-nic.md:167]: What will we put here as the a value? The name of the interface? I would assume only cluster-manager can update this so it would need to determine the value from the host-addresses annotation?

File: enhancements/network/egress-ip-multi-nic.md [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+
+This enhancement is proposing enhancing cluster manager if the egress IP is within a subnet of a known non-primary network, then it
+will also consider which node hosts this secondary network.
+
+Cluster manager will therefore need to understand all available networks within the cluster and this enhancement proposes
+to add all non-ovn managed interfaces addresses and subnet mask to existing annotation `k8s.ovn.org/host-addresses` which is
+added to all node custom resource objects.
+
+Egress IP must not equal an address already assigned to an interface, and we can use the aforementioned `k8s.ovn.org/host-addresses`
+annotation to detect this scenario. The addresses that can be added into this annotation are filtered by the following criteria:
```
Comment (trozet) [inline enhancements/network/egress-ip-multi-nic.md:57]: just to make sure, we are not configuring egress IPs on the interface right?

File: enhancements/network/egress-ip-multi-nic.md [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+- Egress IP interface is now a child interface and attached to a bond which is now a different IP of the original EIP interface. Reconcile to move the EIP
+to a different network and if none found, clear EIP node status and let cluster manager reconcile and find a new node
+- Egress IP balancing of non-ovn EIP assignments is the same as current EIP balancing
+- Mixing Egress IPs from ovn and non-ovn managed networks
+- GARPs are sent when egress IP moves from a non-ovn managed network
+- Link goes down
+- EIP removed
+- Original network that allowed EIP assignment is removed
+
+### API Extensions
```
Comment (trozet) [inline enhancements/network/egress-ip-multi-nic.md:156]: so from what I understand there is no API change needed to indicate you want to use this new behavior, where we determine which interface suits the IP a user gave us. Is that right? I'm wondering is there an issue here, where a customer has been using egress IP and assuming it will go out their br-ex interface, and once they upgrade to this it could pick a different interface?

If this is a real issue, should we make this enabled via some new API?

File: enhancements/network/egress-ip-multi-nic.md [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+- Selecting the wrong network types from interfaces
+  Only consider addresses with scope `Global`.
+
+[1] https://en.wikipedia.org/wiki/Longest_prefix_match
+
+## Design Details
+### Cluster Manager
+If egress IP is within a subnet of a non-ovn managed network, it will need to assign this egress IP to a node which hosts this network.
+
+### OVN kube controller (aka ovn kube master, aka network controller manager)
```
Comment (trozet) [inline enhancements/network/egress-ip-multi-nic.md:209]: So iirc egress IP with shared and lgw both works in a shared gw type of way today. In other words, OVN SNAT is handling the egress IP and it never goes to the kernel. Will you be changing that behavior so that it always goes to the kernel, or is it conditional on if the egress interface is attached to br-ex or not?

If there is some change in behavior, we need to consider how we will upgrade and remove stale things from the GR and/or LRPs from ovn_cluster_router.

**Review by trozet — COMMENTED** [review]
very well written! Some comments inline.

File: enhancements/network/egress-ip-multi-nic.md (hunk 20 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:167]: It is not the name of an interface but a network (i.e. `10.10.10.4/24`).
It will be a network found in `host-addresses` annotation.
Correct, only CM can set it.
Nothing will depend on this value though to allow safe upgrades.
If any other component needs to calculate the network the can.
This is mainly for debug.

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:57]: Dont think I understand you. Can you rephrase?

File: enhancements/network/egress-ip-multi-nic.md (hunk 22 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:156]: Can you see this actually happening in the field? 
Also, is that a problem if it goes out another interface? Theyll still get their traffic out into the right network but there will be a different MAC in the pkt than what they saw before.
If you think this is an issue, happy to extend the API.

File: enhancements/network/egress-ip-multi-nic.md (hunk 23 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:209]: If the egress IP is still selected to go out br-ex, then it will perform like today and egress on br-ex.

Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:209]: but youre right - what could happen is egress IP pre upgrade was selected to go out br-ex, but after upgrade CM found a "better" network to host it using longest prefix matching so in that scenario we would have to clean up LRPs. I added a test case to the enhancement.

Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:210]: We could add a caveat to CM selection in that if the EIP falls within the ovn network, then dont use longest prefix match, just select that ovn network... WDYT? That would help us not have to make api changes.

Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:210]: Seems like a sensible change to me..

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (trozet) [reply enhancements/network/egress-ip-multi-nic.md:57]: I mean we are not configuring an Egress IP address on any linux interface right? I just wanted to make sure.

File: enhancements/network/egress-ip-multi-nic.md (hunk 22 above)
Comment (trozet) [reply enhancements/network/egress-ip-multi-nic.md:157]: I could see it happening. Maybe we should use a detectNetwork (there is probably a better name you can come up with) field in the spec to determine whether this behavior is on or off?

File: enhancements/network/egress-ip-multi-nic.md (hunk 23 above)
Comment (trozet) [reply enhancements/network/egress-ip-multi-nic.md:210]: Well I'm also asking because im wondering if now is a good time to make egress IP work the right way with local gateway mode. Since you will be basically adding the same logic anyway for multiple nics, we might as well convert local gw to always send to kernel, as it originally should have been.

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:57]: yes we are - whatever interface hosts the egress IP, we add the egress IP to the interface so ARPs work.

File: enhancements/network/egress-ip-multi-nic.md (hunk 22 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:157]: I dont think theres a need for a new field if we first try and always see if we can host the egress IP on the primary network and only after failing to host it there, we consider secondary interfaces.

File: enhancements/network/egress-ip-multi-nic.md (hunk 23 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:210]: yeah - it maybe a good time to do this but I am not committing to it. We can see in the implementation.

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (trozet) [reply enhancements/network/egress-ip-multi-nic.md:57]: OK then how will we reconcile: "Egress IP must not equal an address already assigned to an interface, and we can use the aforementioned `k8s.ovn.org/host-addresses`
annotation to detect this scenario."

Like if we put the IP on the interface, then our node tracker will detect it and add it the list of host-addresses then it == egress ip?

File: enhancements/network/egress-ip-multi-nic.md (hunk 22 above)
Comment (trozet) [reply enhancements/network/egress-ip-multi-nic.md:157]: ok so by try you mean if the primary interface has a subnet that matches the requested egress IP, we prefer that one instead of doing longest prefix match?

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:57]: Node IP today will not consume egress IPs cause they either `/32` or `/128` and I also added a label to each EIP i assign so I can also ensure we track it after reboots/upgrades.

File: enhancements/network/egress-ip-multi-nic.md (hunk 22 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:157]: exactly

File: enhancements/network/egress-ip-multi-nic.md (hunk 21 above)
Comment (martinkennelly) [reply enhancements/network/egress-ip-multi-nic.md:57]: Node IP today will _not_ consume addresses that aren't `/32` or `/128`.

**Review by kyrtapz — APPROVED** [review]
